### PR TITLE
[Flux] Add advanced training script + support textual inversion inference

### PR DESCRIPTION
**Pull request by linoytsaban**
This PR adds an advanced version of the dreambooth lora flux script, with additional accompanying update to FluxPipline: 

-  adds an advanced script with pivotal tuning feature for CLIP & T5 encoder
- - `--train_text_encoder_ti` (enables CLIP pivotal tuning)
- - `--enable_t5_ti` adds T5 to the mix
- - `--train_text_encoder_ti_frac` the portion of epochs to train embeddings on, when using just clip
- - `--train_transformer_frac` the portion of epochs to train the transformer on, train_transformer_frac==0 will trigger a "pure_textual_inversion" train run (= "classical" textual inversion, no optimization of transformer lora layers)
- - `--initializer_token` the token to init the textual inversion embeddings with instead of randomly (random by default).
-  `--lora_blocks` the blocks\layers to apply lora training on. 

-  modifies Flux Pipeline (and related pipelines) to allow for textual inversion inference (that is also required by Loras trained with pivotal tuning)

_motivation:_ to allow for fast iterations & experimental features. I think it'd be good to fork the canonical script and bring the above changes into the advanced folder in a similar manner to what we have for SDXL.

cc @apolinario

### CODE REVIEW

**Comment by linoytsaban** [comment]
@sayakpaul @apolinario wdyt about me closing #9160 and moving the changes here as I suggest above?

**Comment by sayakpaul** [comment]
I would prefer #9160 as it helps to review the changes to the canonical script in isolation. Would that work for you?

**Comment by linoytsaban** [comment]
Yeah that makes sense, how about I adjust #9160 to be only for latent caching support for the canonical script? that way the remain isolated

**Comment by sayakpaul** [comment]
That works for me @linoytsaban!

File: examples/advanced_diffusion_training/README_flux.md [hunk 1]
```diff
@@ -0,0 +1,338 @@
...
+To activate pivotal tuning for both encoders, add the flag `--enable_t5_ti`. 
+* When not fine-tuning the text encoders, we ALWAYS precompute the text embeddings to save memory.
+* **pure textual inversion** - to support the full range from pivotal tuning to textual inversion we introduce `--train_transformer_frac` which controls the amount of epochs the transformer LoRA layers are trained. By default, `--train_transformer_frac==1`, to trigger a textual inversion run set `--train_transformer_frac==0`. Values between 0 and 1 are supported as well, and we welcome the community to experiment w/ different settings and share the results!
+* **token initializer** - similar to the original textual inversion work, you can specify a token of your choosing as the starting point for training. By default, when enabling `--train_text_encoder_ti`, the new inserted tokens are initialized randomly. You can specify a token in `--initializer_token` such that the starting point for the trained embeddings will be the embeddings associated with your chosen `--initializer_token`.
```
Comment (apolinario) [inline examples/advanced_diffusion_training/README_flux.md:84]: Here, if this is not a single token but rather it would take as many tokens as the person initializes, we probably should change the name to `intializer_concept`

File: examples/advanced_diffusion_training/README_flux.md [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+instance_token = "<s0><s1>"
+prompt = f"a {instance_token} icon of an orange llama eating ramen, in the style of {instance_token}"
+
+image = pipe(prompt=prompt, num_inference_steps=25, cross_attention_kwargs={"scale": 1.0}).images[0]
+image.save("llama.png")
+```
+
+### Inference - pure textual inversion
+In this case, we don't load transformer layers as before, since we only optimize the text embeddings. The output of a textual inversion train run is a
+`.safetensors` file containing the trained embeddings for the new tokens either for the CLIP encoder, or for both encoders (CLIP and T5)
```
Comment (apolinario) [inline examples/advanced_diffusion_training/README_flux.md:288]: very nice!

**Comment by linoytsaban** [comment]
@sayakpaul  Yes ofc will add some examples soon! 
It is important to note that some of these are _very_ experimental and are meant for the community to explore and experiment with, with the motivation for all being integrating techniques (or modified versions of them) that showed success in dreambooth lora fine tuning of previous base models

**Comment by apolinario** [comment]
IMO we could merge the current model and add exmaples/samples built both by us and the community on a later stage, as the idea is less about bragging this changes have a big positive effect, and more to invite the community to test out the many possibilities this script brings (T5 fine-tuning, pivotal tuning with only CLIP, pivotal tuning with CLIP+T5, pure textual inversion, few different ways of doing token initialization). I think it's best to descentralize this experiments with the community vs running all ourselves (we're running them anyways though, but probably not needed as a bottleneck)

**Comment by sayakpaul** [comment]
Having the results apriori is tremendously beneficial. We don't have to be exhaustive in our hparam space here just 3/4 should suffice IMO.

**Comment by apolinario** [comment]
The idea of the advanced folder was a bit more loose and experimental and freeform - similar to the `community` folder, so I don't think we need to block merging before experiments, and as mentioned before, we'd like to share the experiments with the community. This said, Linoy is running a very extensive benchmarking of a lot of those params, but I don't see why bottlenecking.

**Comment by sayakpaul** [comment]
Sure but the PR doesn't supplement a single result gained from the additions. My understanding is that we have at least one result that we feel confident about regarding the advancements we're introducing in this PR. Or is that not a reasonable expectation? 

In that case, I think it's totally fine not to wait for even 3/4 of the experimental results before merging (which I still think is reasonable enough as an expectation).

**Comment by apolinario** [comment]
This is a bit different than other stuff: we are adding to the script techniques that are well used and work on SDXL: pivotal tuning, pure textual inversion, t5 text encoder training, and some other minor things. Those are tools on a toolbox and we want to put this tools out and invite the community to try/test/expand; The goal is *not* to provide results that we know that are improved with this additions, but rather that there are tools that have a big combinatorial analysis of possibilities and we want to invite people to test

I honestly don't see an issue with sharing this wht the community and bringing results/what worked/what doesn't on a later stage

**Comment by linoytsaban** [comment]
@sayakpaul Since there's quite a bit going on in this script, it would be great to get your thoughts if you could give the code a look in the meantime 🤗 

As poli mentioned I'm running multiple experiments on different uses cases (style, object, face) with different combinations of the new features so it takes a bit of time (that way we'll have some sort of benchmark of the contributions) but I'll add to this thread soon!

It'd be awesome if we can parallel this with the review as this will also allow for faster iterations in case there will be changes that require us to do more training runs

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py [hunk 3]
```diff
@@ -0,0 +1,2447 @@
...
+    logger.info(
+        f"Running validation... \n Generating {args.num_validation_images} images with prompt:"
+        f" {args.validation_prompt}."
+    )
+    pipeline = pipeline.to(accelerator.device, dtype=torch_dtype)
+    pipeline.set_progress_bar_config(disable=True)
+
+    # run inference
+    generator = torch.Generator(device=accelerator.device).manual_seed(args.seed) if args.seed else None
+    # autocast_ctx = torch.autocast(accelerator.device.type) if not is_final_validation else nullcontext()
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:232]: Let's remove this.

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py [hunk 4]
```diff
@@ -0,0 +1,2447 @@
...
+    text_encoder_one = class_one.from_pretrained(
+        args.pretrained_model_name_or_path, subfolder="text_encoder", revision=args.revision, variant=args.variant
+    )
+    text_encoder_two = class_two.from_pretrained(
+        args.pretrained_model_name_or_path, subfolder="text_encoder_2", revision=args.revision, variant=args.variant
+    )
+    return text_encoder_one, text_encoder_two
+
+
+def log_validation(
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:214]: We have had a couple of issues when running the validation in between Flux LoRA training. So, let's make sure to verify this.

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py [hunk 5]
```diff
@@ -0,0 +1,2447 @@
...
+        if tracker.name == "wandb":
+            tracker.log(
+                {
+                    phase_name: [
+                        wandb.Image(image, caption=f"{i}: {args.validation_prompt}") for i, image in enumerate(images)
+                    ]
+                }
+            )
+
+    del pipeline
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:252]: Let's also garbage collect `gc.collect()`.

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py [hunk 6]
```diff
@@ -0,0 +1,2447 @@
...
+        raise ValueError(
+            "--train_transformer_frac must be == 1 if text_encoder training / textual inversion is not enabled."
+        )
+    if args.train_transformer_frac < 1 and args.train_text_encoder_ti_frac < 1:
+        raise ValueError(
+            "--train_transformer_frac and --train_text_encoder_ti_frac are identical and smaller than 1. "
+            "This contradicts with --max_train_steps, please specify different values or set both to 1."
+        )
+    if args.enable_t5_ti and not args.train_text_encoder_ti:
+        warnings.warn("You need not use --enable_t5_ti without --train_text_encoder_ti.")
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:775]: Let's use `logger.warning()` for warnings to keep things consistent?

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py [hunk 7]
```diff
@@ -0,0 +1,2447 @@
...
+        return self.num_samples
+
+    def __getitem__(self, index):
+        example = {}
+        example["prompt"] = self.prompt
+        example["index"] = index
+        return example
+
+
+def tokenize_prompt(tokenizer, prompt, max_sequence_length, add_special_tokens=False):
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1127]: I am wondering if we should handle the prompt encoding part, leveraging the `encode_prompt()` method of `FluxPipeline`. If we can do this, it will help us in getting rid of a lot of boilerplate code that we're reusing from the pipeline. 

It would look something like this:

```python
text_encoder = CLIPTextModel.from_pretrained(
    ckpt_id, subfolder="text_encoder", torch_dtype=torch.bfloat16
)
text_encoder_2 = T5EncoderModel.from_pretrained(
    ckpt_id, subfolder="text_encoder_2", torch_dtype=torch.bfloat16
)
tokenizer = CLIPTokenizer.from_pretrained(ckpt_id, subfolder="tokenizer")
tokenizer_2 = T5TokenizerFast.from_pretrained(ckpt_id, subfolder="tokenizer_2")

pipeline = FluxPipeline.from_pretrained(
    ckpt_id,
    text_encoder=text_encoder,
    text_encoder_2=text_encoder_2,
    tokenizer=tokenizer,
    tokenizer_2=tokenizer_2,
    transformer=None,
    vae=None,
).to("cuda")

with torch.no_grad():
    print("Encoding prompts.")
    prompt_embeds, pooled_prompt_embeds, text_ids = pipeline.encode_prompt(
        prompt=prompt, prompt_2=None, max_sequence_length=256
    )
```

Perfectly fine if we completely get rid of functions like `load_text_encoders()`. WDYT?

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py [hunk 8]
```diff
@@ -0,0 +1,2447 @@
...
+    ) -> torch.Tensor:
+        ## ref https://github.com/huggingface/diffusers/blob/fbe29c62984c33c6cf9cf7ad120a992fe6d20854/examples/dreambooth/train_dreambooth_sd3.py#L1578
+        ## Add noise according to flow matching.
+        ## zt = (1 - texp) * x + texp * z1
+
+        # sigmas = get_sigmas(timesteps, n_dim=model_input.ndim, dtype=model_input.dtype)
+        # noisy_model_input = (1.0 - sigmas) * model_input + sigmas * noise
+
+        # timestep needs to be in [0, 1], we store them in [0, 1000]
+        # noisy_sample = (1 - timestep) * latent + timestep * noise
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1330]: Let's remove the unnecessary comments?

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py [hunk 9]
```diff
@@ -0,0 +1,2447 @@
...
+        # if args.initializer_concept is provided, we ignore args.num_new_tokens_per_abstraction
+        else:
+            token_ids = tokenizer_one.encode(args.initializer_concept, add_special_tokens=False)
+            num_new_tokens_per_abstraction = len(token_ids)
+            if args.enable_t5_ti:
+                token_ids_t5 = tokenizer_two.encode(args.initializer_concept, add_special_tokens=False)
+                num_new_tokens_per_abstraction = max(len(token_ids), len(token_ids_t5))
+            print(
+                f"initializer_concept: {args.initializer_concept}, num_new_tokens_per_abstraction: {num_new_tokens_per_abstraction}"
+            )
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1527]: We are using a mix of `accelerator.print()`, `logger.info()`, and `print()`. Let's perhaps just stick to one?

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py [hunk 10]
```diff
@@ -0,0 +1,2447 @@
...
+
+    if args.train_text_encoder:
+        text_lora_parameters_one = list(filter(lambda p: p.requires_grad, text_encoder_one.parameters()))
+    # if we use textual inversion, we freeze all parameters except for the token embeddings
+    # in text encoder
+    elif args.train_text_encoder_ti:
+        text_lora_parameters_one = []  # CLIP
+        for name, param in text_encoder_one.named_parameters():
+            if "token_embedding" in name:
+                print("YES 5")
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1711]: Can be removed later?

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py [hunk 11]
```diff
@@ -0,0 +1,2447 @@
...
+    # If no type of tuning is done on the text_encoder and custom instance prompts are NOT
+    # provided (i.e. the --instance_prompt is used for all images), we encode the instance prompt once to avoid
+    # the redundant encoding.
+    if freeze_text_encoder and not train_dataset.custom_instance_prompts:
+        instance_prompt_hidden_states, instance_pooled_prompt_embeds, instance_text_ids = compute_text_embeddings(
+            args.instance_prompt, text_encoders, tokenizers
+        )
+
+    # Handle class prompt for prior-preservation.
+    if args.with_prior_preservation:
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1901]: Just a suggestion. Even though we want to provide options to the community, I think with a model like Flux, the use of prior preservation just really gets diminished, especially when we're doing text encoder fine-tuning as well as textual-inversion.

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py [hunk 12]
```diff
@@ -0,0 +1,2447 @@
...
+
+    vae_config_shift_factor = vae.config.shift_factor
+    vae_config_scaling_factor = vae.config.scaling_factor
+    vae_config_block_out_channels = vae.config.block_out_channels
+    if args.cache_latents:
+        latents_cache = []
+        for batch in tqdm(train_dataloader, desc="Caching latents"):
+            with torch.no_grad():
+                batch["pixel_values"] = batch["pixel_values"].to(
+                    accelerator.device, non_blocking=True, dtype=weight_dtype
```
Comment (sayakpaul) [inline examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1969]: We could also do `pin_mememory=True` for non-blocking transfers. This way, we have a page-locked memory. 

https://discuss.pytorch.org/t/when-to-set-pin-memory-to-true/19723

**Review by sayakpaul — COMMENTED** [review]
Thanks! Most of my comments are pretty minor in nature except for trying to reuse the `encode_prompt()` method on the `FluxPipeline`. LMK if it makes sense.

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py (hunk 7 above)
Comment (linoytsaban) [reply examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1127]: the thing is that for pivotal tuning / textual inversion we need to mess with the tokenizers and text encoders both to add the tokens and access the embeddings and to then optimize so I'm not sure I see how that could work together

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py (hunk 11 above)
Comment (linoytsaban) [reply examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1901]: I tend to agree! my main reason why I would keep it is for multi-concept training where you could utilize prior preservation to maintain the knowledge of a previously _trained_ concept. WDYT? also maybe @apolinario has thoughts on this

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py (hunk 7 above)
Comment (sayakpaul) [reply examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1127]: Okay then disregard.

File: examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py (hunk 11 above)
Comment (sayakpaul) [reply examples/advanced_diffusion_training/train_dreambooth_lora_flux_advanced.py:1901]: Okay if you feel so, we can keep it.

**Comment by sayakpaul** [comment]
@linoytsaban let me know when you need another review.

**Review by sayakpaul — APPROVED** [review]
Thanks!

**Comment by sayakpaul** [comment]
@yiyixuxu could you also give the changes in the pipelines a look? It's just about adding `TextualInversionMixin` so that we can enable pivotal tuning on Flux.

**Review by yiyixuxu — APPROVED** [review]
thanks!

**Review by apolinario — APPROVED** [review]
Let's go! 🚀
